### DISCUSSION TITLE
Enable a work with not yet fetched remote branches.

### DISCUSSION THREAD

**Discussion by iassenev** [gitextensions/gitextensions#9870] [Ideas] [answered]
`git remote show <remote>` lists the branches in the remote with the information if the branch has been fetched already.
It would be very useful to be able to get these branches with the flag <not fetched> in the branches list of the remote with the possibility to fetch such branch(es) individually.

**Comment by pmiossec** [comment]
>git remote show <remote> lists the branches in the remote with the information if the branch has been fetched already.

Can you explain a little more your use case or what you try to achieve because by default, git is not working at the branch level but at the remote level when you do a fetch: all the branches are fetched.

So most of the times **all** the branches are at least fetched once

You also have the possibility to see if remote branches are tracked in the push dialog:

![image](https://user-images.githubusercontent.com/460196/154502648-27707eae-8f23-46ab-92aa-bea5800a049c.png)


>It would be very useful to be able to get these branches with the flag in the branches list of the remote

If we introduce such flag, mostly all the branches will have the flag enabled.
If you don't have a branch because it is new, fetch and you will have it with the flag...

> with the possibility to fetch such branch(es) individually.

And you already have the possibility to fetch it individually (even if I never did it because it a lot easier to launch a "fetch" that will update all the remote branches of the default remote or a "fetch all" that will update all the remote branches of all the remotes):

![image](https://user-images.githubusercontent.com/460196/154501777-9d08061c-08f4-4288-ae37-0eb84873a386.png)

For me, fetch and you will be sure to have fetched all the branches.

So either the request is not clear or there is not a lot of added value....

**Comment by iassenev** [comment]
> Can you explain a little more your use case or what you try to achieve
In our (quite large) repository `fetch all` could take several minutes.
At the same time, the developer usually needs to get the changes in some particular branch and to see the list of the branches and switch to some other branch.

There is a possibility to fetch the current branch only, but one cannot see the list of the not fetched branches although one may do this using the command `git remote show <remote>`.

So, my initial (naive) attempt to introduce this functionality into the GE is the following:
* introduce a possibility to fetch the branches list only in the context menu of the remote (right-click on remote in the remotes tree)
* for the branches, which name has been fetched, but the content hasn't been fetched yet, add some visual differentiation:
** gray color
** append the name with the `(will appear after the fetch)`
** both
** something even better
* introduce a possibility to fetch such branch only

**Reply by mstv** [reply]
(I prefer `fetch all`, too. Sometimes, I fetch a single remote. There's also a background fetch plugin.)

This functionality could be added to the Pull dialog in `fetch` mode.
The list of the `Remote branch` ComboBox could be retrieved from the remote on first drop down. Or else a `Refresh` button could be added right to the ComboBox. Branch decoration is feasible.

![grafik](https://user-images.githubusercontent.com/36601201/154573437-12dabc83-9e46-40df-a893-2a0c5cbe79c7.png)

**Comment by iassenev** [comment]
> This functionality could be added to the Pull dialog in fetch mode.

We have a tree structure of a few hundred branches, it would be difficult to find a branch in such a ComboBox

**Reply by pmiossec** [reply]
@iassenev 

> There is a possibility to fetch the current branch only

yes but that not means that's what you should absolutely do. I would find it painful if every time I will have to find the right branch and fetch only that one.

> but one cannot see the list of the not fetched branches although one may do this using the command `git remote show <remote>`.

What is the added value?

You answered in how you would like to see it implemented but not really on what is the purpose of having this feature.

So not sure if it is a feature request or if you are asking for a confirmation and you would like to implement it yourself...

Can you explain what is your use case and how this feature could improve your workflow (and maybe what is your workflow because it seems counter intuitive based on the default behavior of `git fetch` that is to fetch all the branches).

Because including this feature add, for me, a lot of complexity (and seems not useful for the other users) and maybe some performance problem for something that personally I never, in the more than 10 years of using git, feel the need. That's why I would really  like to understand...
And I also never saw someone rely on the `git remote show <remote>` command for a day to day workflow.

For me the workflow is as simple as fetching every times I need to check the state of my repository versus a remote (and after that your 'not fetched state' will disappear...).

Other difficulties: the output of this command seems more intended for human and it seems very difficult to parse it. I don't know if there is an output format intended for parsing...

**Comment by iassenev** [comment]
The value is in saving time: instead of waiting for 10-15min of fetching all branches, one may fetch branches list instead and switch to some new remote branch in a few seconds.

**Reply by RussKie** [reply]
10-15 minutes? Are on dial-up and how many branches are you fetching? Even in extremely busy repos like Visual Studio with hundreds of branches and 180K tags it takes under a minute to fetch all over a VPN...

**Reply by iassenev** [reply]
> 10-15 minutes? Are on dial-up and how many branches are you fetching? Even in extremely busy repos like Visual Studio with hundreds of branches and 180K tags it takes under a minute to fetch all over a VPN...

A gigabit LAN, a repository with hundreds of branches, a few new branches are usually fetched.

The fetch started to be much slower when the number of entities in the working copy becomes a few million. Most of the time the server spends gathering and compressing the diffs to pass via the network, not the transmit itself.

**Reply by vbjay** [reply]
Might want to see if a git repack or git GC on the server copy (where you are fetching) is possible. See if it can rebuild the commit graph and optimize the packs it has. Also use some of the scripts in 


https://github.com/gitextensions/gitextensions/wiki/Diagnosing-problems

To see if you might want to evaluate what you are sticking in the repo.

**Comment by gerhardol** [comment, marked as answer]
You can view the remote branches when _pushing_ multiple branches.
This can be used before fetching a specific branch.
This uses `ls-remote --heads` that is parsable.
![image](https://user-images.githubusercontent.com/6248932/154753669-726ee989-13a3-46de-afea-a6ba2f6de628.png)

If something is changed, this setting could be used to fetch the remote branches in the Pull/Fetch dialog too.
But it would be a pull-request-only feature

**Reply by mstv** [reply]
The Push dialog even has a `Pull` button. Unfortunately, it does not populate the Pull dialog with the selected branch.
So, you'll need to copy & paste the name. Or contribute the little remaining parts in a PR. 😉

![image](https://user-images.githubusercontent.com/36601201/154758573-8a350193-a068-43e1-8244-5f34bb11f101.png)

**Comment by RussKie** [comment]
This sounds like a server side issue. I don't think it'd be a right
approach to solve this on a client.

In any case the app is a UI layer over the git cli, and you can author
scripts (Settings > Scripts) to fetch required information on demand, and
the app will show it.
